make AlignToVideo sensible with tolerance=0

TheOneric opened the pull request:
Afaik the AlignToVideo feature is specific to wangqr/Aegisub.

Using the “Align to video” feature with a tolerance of 0 would previously always result of the starttime being equal to the start time of the second next frame and similar with endtime, so `start<end` which doesn't make sense as every colour ought to be equal to itself.
This seemed simple enough, so I gave it a try. In addition a typo and a minor cosmetic nit are being fixed with this pr.

It also prevents arbitrary colours from being selected. I don't think it makes much sense to select a colour different from the one at the specified position – if not the logic of the main commit does not hold up as explained in the commit message.
I'm not really familiar with wxWidgets (or any GUI lib for that matter) so I'm not sure if my approach to disable arbitrary colour selection is sensible enough or if there's a better alternative to display colour without a button.

<br />

On an unrelated side note, I don't really understand what the goal of the pixel traversal in all four cardinal directions is. If a large body of similar colour is selected as a source this makes this feature very inperfomant (which is very noticeable for me). But maybe there's a good reason for it? Would a fixed position with a (configurable) position tolerance achieve the same? *(while limiting performance impact)*
*(Also this may lead to premature(?) abortion if the selected pixel still has the original value±tolerance, but the `l`,`r`,`d`,`u` positions move closer to the selected pixel.)* 

**EDIT:** force pushed because I initially forgot to also update the `aegisub.pot` file along with the `*.po` files.

TheOneric committed:
Readme: Fix Typo

TheOneric committed:
dialog_align: Remove point at the end of the tooltips

No other tooltips end with points, even if they're a sentence.These
ones aren't even sentences.
Also updates all *.po files accordingly, with an hacked in
POT-Creation-Date due to issues with make_pot.sh.

TheOneric committed:
AlignToVideo: Handle tolerance = 0 correctly

Checking if the diff is '>' instead of '>=' will always fail if
tolerance=0, even if the colours are identical. This will cause the line
to get a startime greter than its end time, which is not desireable.

Rounding errors and limits of floating type precision might still affect
the comparison. An additional sanity check before calculation is added
to ensure the selected position and colour match within tolerance.
This allows us to refactor the search code to never check the starting
frame and guanrantees valid timings with start<end even with
rounding errors.

wangqr commented:
For the tooltip part, you only need to manually edit the cpp code, and generate the translation file by running `./make_pot.sh` in `po/`.

This way the `POT-Creation-Date` will be updated properly.

wangqr commented:
The original use case for AlignToVideo is for videos that already have hard subs (subtitles that baked into video). This tool can quickly set the start / end time of the current ASS line to the same as the hard sub in original video when doing subtitle translation. So based on this, I'll explain some of your questions.

> I don't think it makes much sense to select a colour different from the one at the specified position ...

Usually we know the exact color used in hard subs beforehand. This is because most video author won't change the font color used in their videos. Unlike retriving the color from current frame, manually setting the color will not be affected by video compression artifacts, thus can be used with a smaller tolerance value to be more accurate.

It's good that you brought this up. We may add some improvements here:
* Add a sanity check that the color in current frame is within tolerance of the specified color,
* Add a way (right click?) to allow user pick a position without changing the reference color.

> I don't really understand what the goal of the pixel traversal in all four cardinal directions is.

It was added in 9ba1f90e75df1cb578d86948b8ba3c5c861dd574 . For example we have 2 consecutive subtitle lines with no gap in between. They share the same font color. Now we want to set the start / end time to be the same as first subtitle line. Without traversal, we need to pick a pixel that is in the glyph for first subtitle line, but not in glyph for second. Otherwise if the pixel is in glyph for both lines, the end time will be set to the end time of the second subtitle line. Now with traversal, as long as the characters at the location we pick are different across 2 lines, the second line will not be able to pass the test, and the end time will be set to the end time of the first subtitle line.

I have not profiled the case if a large body of similar color is selected. But in the hard sub use case, most CPU time will be used for video decoding (`provider->GetFrame()`). Performance impact of intra-frame tests are negligible.

TheOneric committed:
AlignToVideo: Improve rgb->lab precision

TheOneric commented:
> It [the pixel traversal] was added in 9ba1f90 . […]

Thanks for the explanation.


> Usually we know the exact color used in hard subs beforehand. This is because most video author won't change the font color used in their videos. Unlike retriving the color from current frame, manually setting the color will not be affected by video compression artifacts, thus can be used with a smaller tolerance value to be more accurate.
>
>It's good that you brought this up. We may add some improvements here:
>
>  - Add a sanity check that the color in current frame is within tolerance of the specified color,
 > - Add a way (right click?) to allow user pick a position without changing the reference color.

I allowed arbitrary colours again and instead added a sanity check as you suggested. I couldn't figure out how to create an "update only-position"-event, but it seems like it would be useful.

<br />

Due to the mentioned issues with `make_pot.sh` The newly added string, is not yet included in the pot and po files. Running it twice still strips some `c-format` comments from old strings, and entries in the `aegisub.pot` file are still reorder. Though I don't understand why. See:
```diff
+#: ../automation/autoload/strip-tags.lua:17
+msgid "Strip tags"
+msgstr ""
+
+#: ../automation/autoload/strip-tags.lua:18
+msgid "Remove all override tags from selected lines"
+msgstr ""
+
+#: ../automation/autoload/strip-tags.lua:28
+msgid "strip tags"
+msgstr ""
+
 #: ../automation/autoload/cleantags-autoload.lua:31
 msgid "Clean Tags"
 msgstr ""
@@ -6321,14 +6337,6 @@ msgid ""
 "information on how to use this."
 msgstr ""

-#: ../automation/autoload/karaoke-auto-leadin.lua:32
-msgid "Automatic karaoke lead-in"
-msgstr ""
-
-#: ../automation/autoload/karaoke-auto-leadin.lua:33
-msgid "Join up the ends of selected lines and add \\\\k tags to shift karaoke"
-msgstr ""
-
 #: ../automation/autoload/macro-1-edgeblur.lua:6
 msgid "Add edgeblur"
 msgstr ""
@@ -6349,16 +6357,12 @@ msgstr ""
 msgid "Convert Latin letters to SJIS fullwidth letters"
 msgstr ""

-#: ../automation/autoload/strip-tags.lua:17
-msgid "Strip tags"
-msgstr ""
-
-#: ../automation/autoload/strip-tags.lua:18
-msgid "Remove all override tags from selected lines"
+#: ../automation/autoload/karaoke-auto-leadin.lua:32
+msgid "Automatic karaoke lead-in"
```

<br />

Partially this can be solved with the following patch, which will also make it more portable and safe *(`echo -e` is not POSIX)*.
This will do some resorting in `aegisub.pot` initially on strings from the same source line, but should ensure no-reodering in the future. (still leaves the problem of old strings in `*.po`-files to be solved):
```diff
diff --git a/po/make_pot.sh b/po/make_pot.sh
index 08d4feb16..ebf0215e0 100755
--- a/po/make_pot.sh
+++ b/po/make_pot.sh
@@ -1,5 +1,8 @@
 #!/bin/sh
 
+#Abort on error
+set -e
+
 maybe_append() {
   while read -r msg; do
     msgfile=$(echo $msg | cut -d'|' -f1)
@@ -7,13 +10,15 @@ maybe_append() {
     msgid=$(echo $msg | cut -d'|' -f3-)
 
     if ! grep -Fq "msgid $msgid" aegisub.pot; then
-      echo -e "\n#: $msgfile:$msgline\nmsgid $msgid\nmsgstr \"\"\n" >> aegisub.pot
+      printf "\n#: %s:%s\nmsgid %s\nmsgstr \"\"\n\n" \
+             "$msgfile" "$msgline" "$msgid" >> aegisub.pot
     fi
   done
 }
 
-find ../src -name \*.cpp -o -name \*.h | LC_ALL=C sort \
-  | xgettext --files-from=- -o - --c++ -k_ -kSTR_MENU -kSTR_DISP -kSTR_HELP -kfmt_tl -kfmt_plural:2,3 -kwxT -kwxPLURAL:1,2 \
+find ../src -name \*.cpp -o -name \*.h \
+  | xgettext --files-from=- -o - --c++ --sort-by-file \
+             -k_ -kSTR_MENU -kSTR_DISP -kSTR_HELP -kfmt_tl -kfmt_plural:2,3 -kwxT -kwxPLURAL:1,2 \
   | sed 's/SOME DESCRIPTIVE TITLE./Aegisub 3.3+/' \
   | sed 's/YEAR/2005-2020/' \
   | sed "s/THE PACKAGE'S COPYRIGHT HOLDER/Rodrigo Braz Monteiro, Niels Martin Hansen, Thomas Goyne et. al./" \

```

wangqr commented:
Oh I see that the reorder happens in lua files, which is not sorted in make_pot.sh

https://github.com/wangqr/Aegisub/blob/dev/po/make_pot.sh#L36

wangqr commented:
In case you've missed the comment in `src/dialog_align.cpp ` above: I'm not sure about the
```C++
		while (pos >= 0)
		{
			pos -= 2;
			if (!CHECK_EXISTS_POS)
				break;
		}
```
Isn't this allowing `pos` to be negative when calling `check_exists`? I think it should be
```C++
		do {
			pos -= 2;
		} while (pos >= 0 && CHECK_EXISTS_POS);
```
Correct me if I'm wrong

TheOneric committed:
po/make_pot: Fix sorting and make POSIX compliant

Replaces non POSIX `echo -e` with an equivalent printf call and
fixes sorting of lua files.

This still leaves a problem with the order of old commented strings
being being reversed and there #, c-format comments being stripped.
See: https://savannah.gnu.org/bugs/?58778

TheOneric commented:
> Isn't this allowing pos to be negative when calling check_exists?

Ah, yes you're right, thanks. It's only being clamped after the looop exits. Changed both loops.
Also included a patch for `make_pot.sh`. Due to the old commented lines being still stripped of their `#, c-format`-lines (when invoking the script twice), I still did not include the new string in the pot and po files.
